<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Records · XAM.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="XAM.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">XAM.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><span class="tocitem">Home</span><ul><li><a class="tocitem" href="../../">Overview</a></li><li class="is-active"><a class="tocitem" href>Records</a><ul class="internal"><li><a class="tocitem" href="#SAM-and-BAM-Headers"><span>SAM and BAM Headers</span></a></li><li><a class="tocitem" href="#Accessing-auxiliary-data"><span>Accessing auxiliary data</span></a></li></ul></li><li><a class="tocitem" href="../files/">File I/O</a></li></ul></li><li><a class="tocitem" href="../sam/">SAM</a></li><li><a class="tocitem" href="../bam/">BAM</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Home</a></li><li class="is-active"><a href>Records</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Records</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/XAM.jl/blob/develop/docs/src/man/records.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><p>The <code>XAM</code> package offers high-performance tools for SAM and BAM file formats, which are the most popular file formats.</p><p>If you have questions about the SAM and BAM formats or any of the terminology used when discussing these formats, see the published <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">specification</a>, which is maintained by the <a href="https://samtools.github.io/">samtools group</a>.</p><p>A very very simple SAM file looks like the following:</p><pre><code class="nohighlight hljs">@HD VN:1.6 SO:coordinate
@SQ SN:ref LN:45
r001   99 ref  7 30 8M2I4M1D3M = 37  39 TTAGATAAAGGATACTG *
r002    0 ref  9 30 3S6M1P1I4M *  0   0 AAAAGATAAGGATA    *
r003    0 ref  9 30 5S6M       *  0   0 GCCTAAGCTAA       * SA:Z:ref,29,-,6H5M,17,0;
r004    0 ref 16 30 6M14N5M    *  0   0 ATAGCTTCAGC       *
r003 2064 ref 29 17 6H5M       *  0   0 TAGGC             * SA:Z:ref,9,+,5S6M,30,1;
r001  147 ref 37 30 9M         =  7 -39 CAGCGGCAT         * NM:i:1</code></pre><p>Where the first two lines are part of the &quot;header&quot;, and the following lines are &quot;records&quot;. Each record describes how a read aligns to some reference sequence. Sometimes one record describes one read, but there are other cases like chimeric reads and split alignments, where multiple records apply to one read. In the example above, <code>r003</code> is a chimeric read, and <code>r004</code> is a split alignment, and <code>r001</code> are mate pair reads. Again, we refer you to the official <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">specification</a> for more details.</p><p>A BAM file stores this same information but in a binary and compressible format that does not make for pretty printing here!</p><h2 id="SAM-and-BAM-Headers"><a class="docs-heading-anchor" href="#SAM-and-BAM-Headers">SAM and BAM Headers</a><a id="SAM-and-BAM-Headers-1"></a><a class="docs-heading-anchor-permalink" href="#SAM-and-BAM-Headers" title="Permalink"></a></h2><p>Both <code>SAM.Reader</code> and <code>BAM.Reader</code> implement the <code>header</code> function, which returns a <code>SAM.Header</code> object. To extract certain information out of the headers, you can use the <code>find</code> method on the header to extract information according to SAM/BAM tag. Again we refer you to the <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">specification</a> for full details of all the different tags that can occur in headers, and what they mean.</p><p>Below is an example of extracting all the info about the reference sequences from the BAM header. In SAM/BAM, any description of a reference sequence is stored in the header, under a tag denoted <code>SQ</code> (think <code>reference SeQuence</code>!).</p><pre><code class="language-jlcon hljs">julia&gt; reader = open(SAM.Reader, &quot;data.sam&quot;);

julia&gt; find(header(reader), &quot;SQ&quot;)
7-element Array{Bio.Align.SAM.MetaInfo,1}:
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr1 LN=30427671
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr2 LN=19698289
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr3 LN=23459830
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr4 LN=18585056
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr5 LN=26975502
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=chloroplast LN=154478
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=mitochondria LN=366924
</code></pre><p>In the above we can see there were 7 sequences in the reference: 5 chromosomes, one chloroplast sequence, and one mitochondrial sequence.</p><h2 id="Accessing-auxiliary-data"><a class="docs-heading-anchor" href="#Accessing-auxiliary-data">Accessing auxiliary data</a><a id="Accessing-auxiliary-data-1"></a><a class="docs-heading-anchor-permalink" href="#Accessing-auxiliary-data" title="Permalink"></a></h2><p>SAM and BAM records support the storing of optional data fields associated with tags.</p><p>Tagged auxiliary data follows a format of <code>TAG:TYPE:VALUE</code>. <code>TAG</code> is a two-letter string, and each tag can only appear once per record. <code>TYPE</code> is a single case-sensetive letter which defined the format of <code>VALUE</code>.</p><table><tr><th style="text-align: right">Type</th><th style="text-align: right">Description</th></tr><tr><td style="text-align: right">&#39;A&#39;</td><td style="text-align: right">Printable character</td></tr><tr><td style="text-align: right">&#39;i&#39;</td><td style="text-align: right">Signed integer</td></tr><tr><td style="text-align: right">&#39;f&#39;</td><td style="text-align: right">Single-precision floating number</td></tr><tr><td style="text-align: right">&#39;Z&#39;</td><td style="text-align: right">Printable string, including space</td></tr><tr><td style="text-align: right">&#39;H&#39;</td><td style="text-align: right">Byte array in Hex format</td></tr><tr><td style="text-align: right">&#39;B&#39;</td><td style="text-align: right">Integer of numeric array</td></tr></table><p>For more information about these tags and their types we refer you to the <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAM/BAM specification</a> and the additional <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">optional fields specification</a> document.</p><p>There are some tags that are reserved, predefined standard tags, for specific uses.</p><p>To access optional fields stored in tags, you use <code>getindex</code> indexing syntax on the record object. Note that accessing optional tag fields will result in type instability in Julia. This is because the type of the optional data is not known until run-time, as the tag is being read. This can have a significant impact on performance. To limit this, if the user knows the type of a value in advance, specifying it as a type annotation will alleviate the problem:</p><p>Below is an example of looping over records in a bam file and using indexing syntax to get the data stored in the &quot;NM&quot; tag. Note the <code>UInt8</code> type assertion to alleviate type instability.</p><pre><code class="language-julia hljs">for record in open(BAM.Reader, &quot;data.bam&quot;)
    nm = record[&quot;NM&quot;]::UInt8
    # do something
end</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Overview</a><a class="docs-footer-nextpage" href="../files/">File I/O »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 28 February 2023 01:44">Tuesday 28 February 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
