<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>File I/O · XAM.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="XAM.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">XAM.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><span class="tocitem">Home</span><ul><li><a class="tocitem" href="../../">Overview</a></li><li><a class="tocitem" href="../records/">Records</a></li><li class="is-active"><a class="tocitem" href>File I/O</a><ul class="internal"><li><a class="tocitem" href="#Reading-SAM-and-BAM-files"><span>Reading SAM and BAM files</span></a></li><li><a class="tocitem" href="#Writing-SAM-and-BAM-files"><span>Writing SAM and BAM files</span></a></li><li><a class="tocitem" href="#Getting-records-overlapping-genomic-features"><span>Getting records overlapping genomic features</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../sam/">SAM</a></li><li><a class="tocitem" href="../bam/">BAM</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Home</a></li><li class="is-active"><a href>File I/O</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>File I/O</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/XAM.jl/blob/develop/docs/src/man/files.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Readers-and-writers"><a class="docs-heading-anchor" href="#Readers-and-writers">Readers and writers</a><a id="Readers-and-writers-1"></a><a class="docs-heading-anchor-permalink" href="#Readers-and-writers" title="Permalink"></a></h1><h2 id="Reading-SAM-and-BAM-files"><a class="docs-heading-anchor" href="#Reading-SAM-and-BAM-files">Reading SAM and BAM files</a><a id="Reading-SAM-and-BAM-files-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-SAM-and-BAM-files" title="Permalink"></a></h2><p>Below is an example of a typical script iterating over all records in a file:</p><pre><code class="language-julia hljs">using XAM

# Open a BAM file.
reader = open(BAM.Reader, &quot;data.bam&quot;)

# Iterate over BAM records.
for record in reader
    # `record` is a BAM.Record object.
    if BAM.ismapped(record)
        # Print the mapped position.
        println(BAM.refname(record), &#39;:&#39;, BAM.position(record))
    end
end

# Close the BAM file.
close(reader)</code></pre><p>The size of a BAM file is often extremely large. The iterator interface demonstrated above allocates an object for each record and that may be a bottleneck of reading data from a BAM file. In-place reading reuses a pre-allocated object for every record and less memory allocation happens in reading:</p><pre><code class="language-julia hljs">reader = open(BAM.Reader, &quot;data.bam&quot;)
record = BAM.Record()
while !eof(reader)
    empty!(record)
    read!(reader, record)
    # do something
end</code></pre><h2 id="Writing-SAM-and-BAM-files"><a class="docs-heading-anchor" href="#Writing-SAM-and-BAM-files">Writing SAM and BAM files</a><a id="Writing-SAM-and-BAM-files-1"></a><a class="docs-heading-anchor-permalink" href="#Writing-SAM-and-BAM-files" title="Permalink"></a></h2><p>In order to write a BAM or SAM file, you must first create a <code>SAM.Header</code>.</p><p>A <code>SAM.Header</code> is constructed from a vector of <code>SAM.MetaInfo</code> objects.</p><p>For example, to create the following simple header:</p><pre><code class="nohighlight hljs">@HD VN:1.6 SO:coordinate
@SQ SN:ref LN:45</code></pre><pre><code class="language-julia hljs">julia&gt; a = SAM.MetaInfo(&quot;HD&quot;, [&quot;VN&quot; =&gt; 1.6, &quot;SO&quot; =&gt; &quot;coordinate&quot;])
SAM.MetaInfo:
    tag: HD
  value: VN=1.6 SO=coordinate

julia&gt; b = SAM.MetaInfo(&quot;SQ&quot;, [&quot;SN&quot; =&gt; &quot;ref&quot;, &quot;LN&quot; =&gt; 45])
SAM.MetaInfo:
    tag: SQ
  value: SN=ref LN=45

julia&gt; h = SAM.Header([a, b])
SAM.Header(SAM.MetaInfo[SAM.MetaInfo:
    tag: HD
  value: VN=1.6 SO=coordinate, SAM.MetaInfo:
    tag: SQ
  value: SN=ref LN=45])
</code></pre><p>Then to create the writer for a SAM file, construct a <code>SAM.Writer</code> using the header and an <code>IO</code> type:</p><pre><code class="language-julia hljs">julia&gt; samw = SAM.Writer(open(&quot;my-data.sam&quot;, &quot;w&quot;), h)
SAM.Writer(IOStream(&lt;file my-data.sam&gt;))
</code></pre><p>To make a BAM Writer is slightly different, as you need to use a specific stream type from the <a href="https://github.com/BioJulia/BGZFStreams.jl">https://github.com/BioJulia/BGZFStreams.jl</a> package:</p><pre><code class="language-julia hljs">julia&gt; using BGZFStreams

julia&gt; bamw = BAM.Writer(BGZFStream(open(&quot;my-data.bam&quot;, &quot;w&quot;), &quot;w&quot;))
BAM.Writer(BGZFStreams.BGZFStream{IOStream}(&lt;mode=write&gt;))
</code></pre><p>Once you have a BAM or SAM writer, you can use the <code>write</code> method to write <code>BAM.Record</code>s or <code>SAM.Record</code>s to file:</p><pre><code class="language-julia hljs">julia&gt; write(bamw, rec) # Here rec is a `BAM.Record`
330780</code></pre><h2 id="Getting-records-overlapping-genomic-features"><a class="docs-heading-anchor" href="#Getting-records-overlapping-genomic-features">Getting records overlapping genomic features</a><a id="Getting-records-overlapping-genomic-features-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-records-overlapping-genomic-features" title="Permalink"></a></h2><p>The <code>eachoverlap</code> method also accepts the <code>Interval</code> type defined in <a href="https://github.com/BioJulia/GenomicFeatures.jl">GenomicFeatures.jl</a>.</p><p>This allows you to do things like first read in the genomic features from a GFF3 file, and then for each feature, iterate over all the BAM records that overlap with that feature.</p><pre><code class="language-julia hljs">using GenomicFeatures
using GFF3
using XAM

# Load genomic features from a GFF3 file.
features = open(collect, GFF3.Reader, &quot;TAIR10_GFF3_genes.gff&quot;)

# Keep mRNA features.
filter!(x -&gt; GFF3.featuretype(x) == &quot;mRNA&quot;, features)

# Open a BAM file and iterate over records overlapping mRNA transcripts.
reader = open(BAM.Reader, &quot;SRR1238088.sort.bam&quot;)
for feature in features
    for record in eachoverlap(reader, feature)
        # `record` overlaps `feature`.
        # ...
    end
end
close(reader)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../records/">« Records</a><a class="docs-footer-nextpage" href="../sam/">SAM »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 28 February 2023 01:44">Tuesday 28 February 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
